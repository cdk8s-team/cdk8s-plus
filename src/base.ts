import { ApiObjectMetadata, ApiObject, ApiObjectMetadataDefinition } from 'cdk8s';
import { Construct } from 'constructs';
import { IApiResource, IApiEndpoint } from './api-resource.generated';
import * as r from './role';
import * as rb from './role-binding';

/**
 * Initialization properties for resources.
 */
export interface ResourceProps {
  /**
   * Metadata that all persisted resources must have, which includes all objects
   * users must create.
   */
  readonly metadata?: ApiObjectMetadata;
}

/**
 * Represents a resource.
 */
export interface IResource {
  /**
   * The Kubernetes name of this resource.
   */
  readonly name: string;

  /**
   * The object's API version (e.g. "authorization.k8s.io/v1")
   */
  readonly apiVersion: string;

  /**
   * The group portion of the API version (e.g. "authorization.k8s.io")
   */
  readonly apiGroup: string;

  /**
   * The object kind (e.g. "Deployment").
   */
  readonly kind: string;
}

/**
 * Base class for all Kubernetes objects in stdk8s. Represents a single
 * resource.
 */
export abstract class Resource extends Construct implements IResource, IApiResource, IApiEndpoint {

  /**
   * The underlying cdk8s API object.
   */
  protected abstract readonly apiObject: ApiObject;

  public get metadata(): ApiObjectMetadataDefinition {
    return this.apiObject.metadata;
  }

  /**
   * The name of this API object.
   */
  public get name(): string {
    return this.apiObject.name;
  }

  /**
   * The object's API version (e.g. "authorization.k8s.io/v1")
   */
  public get apiVersion(): string {
    return this.apiObject.apiVersion;
  }

  /**
   * The group portion of the API version (e.g. "authorization.k8s.io").
   */
  public get apiGroup(): string {
    return this.apiObject.apiGroup;
  }

  /**
   * The object kind (e.g. "Deployment").
   */
  public get kind(): string {
    return this.apiObject.kind;
  }

  public get resourceName(): string | undefined {
    return this.name;
  }

  public readonly abstract resourceType: string;

  public asApiResource(): IApiResource | undefined {
    return this;
  }

  public asNonApiResource(): string | undefined {
    return undefined;
  }

  /**
   * Grants the list of subjects permissions to read this resource.
   */
  public grantRead(...subjects: rb.ISubject[]): rb.RoleBinding {
    const role = new r.Role(this, 'Role', {
      metadata: { namespace: this.metadata.namespace },
    });
    role.allowRead(this);
    return role.bind(...subjects);
  }

  /**
   * Grants the list of subjects permissions to read and write this resource.
   */
  public grantReadWrite(...subjects: rb.ISubject[]): rb.RoleBinding {
    const role = new r.Role(this, 'Role', {
      metadata: { namespace: this.metadata.namespace },
    });
    role.allowReadWrite(this);
    return role.bind(...subjects);
  }

}