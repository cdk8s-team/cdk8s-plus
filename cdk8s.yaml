output: dist
pluginsDirectory: /home/runner/.cdk8s/plugins
imports:
  - k8s@1.27.0
  - k8s@1.28.0
