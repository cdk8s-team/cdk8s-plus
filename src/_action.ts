import { Container } from './container';
import * as k8s from './imports/k8s';

/**
 * Utility class to implement the conversion between our API and the k8s action
 * structure. Used both for probes and handlers.
 *
 * @internal
 */
export class Action {

  public static fromTcpSocket(container: Container, options: { port?: number; host?: string } = {}): k8s.TcpSocketAction {
    return {
      port: k8s.IntOrString.fromNumber(options.port ?? container.port ?? 80),
      host: options.host,
    };
  }

  public static fromCommand(command: string[]): k8s.ExecAction {
    return { command };
  }

  public static fromHttpGet(container: Container, path: string, options: { port?: number } = { }): k8s.HttpGetAction {
    return {
      path,
      port: k8s.IntOrString.fromNumber(options.port ?? container.port ?? 80),
    };
  }
}