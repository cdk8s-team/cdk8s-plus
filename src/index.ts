export * from './base';
export * from './config-map';
export * from './container';
export * from './deployment';
export * from './job';
export * from './pod';
export * from './secret';
export * from './service-account';
export * from './service';
export * from './stateful-set';
export * from './volume';
export * from './ingress';
export * from './probe';
export * from './pvc';
export * from './pv';
export * from './handler';
export * from './workload';
export * from './daemon-set';
export * from './role';
export * from './role-binding';
export * from './namespace';

export * from './api-resource.generated';
