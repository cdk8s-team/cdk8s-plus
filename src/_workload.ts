import { ApiObjectMetadata, ApiObjectMetadataDefinition } from 'cdk8s';
import { Construct } from 'constructs';
import { Resource, ResourceProps } from './base';
import { Container, ContainerProps } from './container';
import { HostAlias, PodSecurityContext, PodSpec, PodSpecProps, RestartPolicy } from './pod';
import { IServiceAccount } from './service-account';
import { Volume } from './volume';

/**
 * Properties for `Workload`.
 */
export interface WorkloadProps extends ResourceProps, PodSpecProps {

  /**
   * The pod metadata of this workload.
   */
  readonly podMetadata?: ApiObjectMetadata;

}

export abstract class Workload extends Resource {

  protected readonly podSpec: PodSpec;

  /**
   * The metadata of pods in this deployment.
   */
  public readonly podMetadata: ApiObjectMetadataDefinition;

  constructor(scope: Construct, id: string, props: WorkloadProps = {}) {
    super(scope, id);

    this.podSpec = new PodSpec(props);
    this.podMetadata = new ApiObjectMetadataDefinition(props.podMetadata);
  }

  public get containers(): Container[] {
    return this.podSpec.containers;
  }

  public get initContainers(): Container[] {
    return this.podSpec.initContainers;
  }

  public get volumes(): Volume[] {
    return this.podSpec.volumes;
  }

  public get restartPolicy(): RestartPolicy | undefined {
    return this.podSpec.restartPolicy;
  }

  public get serviceAccount(): IServiceAccount | undefined {
    return this.podSpec.serviceAccount;
  }

  public get securityContext(): PodSecurityContext {
    return this.podSpec.securityContext;
  }

  public get hostAliases(): HostAlias[] {
    return this.podSpec.hostAliases;
  }

  public addContainer(container: ContainerProps): Container {
    return this.podSpec.addContainer(container);
  }

  public addInitContainer(container: ContainerProps): Container {
    return this.podSpec.addInitContainer(container);
  }

  public addVolume(volume: Volume): void {
    return this.podSpec.addVolume(volume);
  }

  public addHostAlias(hostAlias: HostAlias): void {
    return this.podSpec.addHostAlias(hostAlias);
  }

}