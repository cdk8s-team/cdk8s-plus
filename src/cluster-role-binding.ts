import { ApiObject, Lazy } from 'cdk8s';
import { Construct } from 'constructs';
import { Resource, ResourceProps } from './base';
import * as clusterrole from './cluster-role';
import * as k8s from './imports/k8s';
import * as rb from './role-binding';
import { filterUndefined } from './utils';

/**
 * Properties for `ClusterRoleBinding`.
 */
export interface ClusterRoleBindingProps extends ResourceProps {
  /**
   * The role to bind to.
   */
  readonly role: clusterrole.IClusterRole;
}

/**
 * A ClusterRoleBinding grants permissions cluster-wide to a user or
 * set of users.
 */
export class ClusterRoleBinding extends Resource {
  /**
   * @see base.Resource.apiObject
   */
  protected readonly apiObject: ApiObject;

  public readonly resourceType = 'clusterrolebindings';

  public readonly role: clusterrole.IClusterRole;

  private readonly _subjects: Array<rb.ISubject>;

  constructor(scope: Construct, id: string, props: ClusterRoleBindingProps) {
    super(scope, id);

    this.role = props.role;

    this._subjects = new Array<rb.ISubject>();

    this.apiObject = new k8s.KubeClusterRoleBinding(this, 'Resource', {
      metadata: props.metadata,
      subjects: Lazy.any({ produce: () => this.synthesizeSubjects() }),
      roleRef: Lazy.any({ produce: () => this.synthesizeRoleRef() }),
    });
  }

  /**
   * Adds a subject to the role.
   * @param subjects The subjects to add
   */
  public addSubjects(...subjects: rb.ISubject[]): void {
    for (const subject of subjects) {
      this._subjects.push(subject);
    }
  }

  public get subjects() {
    return [...this._subjects];
  }

  private synthesizeRoleRef(): k8s.RoleRef {
    return {
      apiGroup: this.role.apiGroup,
      kind: this.role.kind,
      name: this.role.name,
    };
  }

  private synthesizeSubjects(): k8s.Subject[] {
    return this._subjects.map((subject) => filterUndefined({
      apiGroup: subject.apiGroup === 'core' ? '' : subject.apiGroup,
      kind: subject.kind,
      name: subject.name,
      namespace: subject.namespace,
    }));
  }
}
