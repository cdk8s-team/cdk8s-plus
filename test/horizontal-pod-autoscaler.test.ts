import { Testing, ApiObject, Duration } from 'cdk8s';
import { Node } from 'constructs';
import * as kplus from '../src';

test('defaultChild', () => {
  const chart = Testing.chart();
  const deployment = new kplus.Deployment(chart, 'Deployment', { containers: [{ image: 'pod' }] });

  const hpa = new kplus.HorizontalPodAutoscaler(chart, 'Hpa', {
    target: deployment,
    maxReplicas: 10,
  });

  const defaultChild = Node.of(hpa).defaultChild as ApiObject;
  expect(defaultChild.kind).toEqual('HorizontalPodAutoscaler');
});

test('default configuration', () => {
  const chart = Testing.chart();
  const deployment = new kplus.Deployment(chart, 'Deployment', { containers: [{ image: 'pod' }] });

  new kplus.HorizontalPodAutoscaler(chart, 'Hpa', {
    target: deployment,
    maxReplicas: 10,
  });

  const manifest = Testing.synth(chart);
  expect(manifest).toMatchSnapshot();
  const spec = manifest[1].spec;
  expect(spec.maxReplicas).toEqual(10);
  expect(spec.minReplicas).toEqual(1);
  expect(spec.metrics).toEqual(undefined);
  expect(spec.behavior.scaleUp).toEqual({
    policies: [
      {
        periodSeconds: 60,
        type: 'Pods',
        value: 4,
      },
      {
        periodSeconds: 60,
        type: 'Percent',
        value: 200,
      },
    ],
    selectPolicy: 'Max',
    stabilizationWindowSeconds: 0,
  });
  expect(spec.behavior.scaleDown).toEqual({
    policies: [
      { periodSeconds: 300, type: 'Pods', value: 1 },
    ],
    selectPolicy: 'Max',
    stabilizationWindowSeconds: 300,
  });
});

test('throws error, when resource container does not have a request/limit defined and metrics is undefined', () => {
  const chart = Testing.chart();
  const deployment = new kplus.Deployment(chart, 'Deployment', { containers: [{ image: 'pod' }] });

  new kplus.HorizontalPodAutoscaler(chart, 'Hpa', {
    target: deployment,
    maxReplicas: 10,
  });

  const manifest = Testing.synth(chart);
  expect(manifest).toMatchSnapshot();
  const spec = manifest[1].spec;
  expect(spec.maxReplicas).toEqual(10);
  expect(spec.minReplicas).toEqual(1);
  expect(spec.metrics).toEqual(undefined);
  expect(spec.behavior.scaleUp).toEqual({
    policies: [
      {
        periodSeconds: 60,
        type: 'Pods',
        value: 4,
      },
      {
        periodSeconds: 60,
        type: 'Percent',
        value: 200,
      },
    ],
    selectPolicy: 'Max',
    stabilizationWindowSeconds: 0,
  });
  expect(spec.behavior.scaleDown).toEqual({
    policies: [
      { periodSeconds: 300, type: 'Pods', value: 1 },
    ],
    selectPolicy: 'Max',
    stabilizationWindowSeconds: 300,
  });
});

test('creates HPA, metrics, scaleUp,and scaleDown are configured', () => {
  const chart = Testing.chart();
  const deployment = new kplus.Deployment(chart, 'Deployment', { containers: [{ image: 'pod' }] });

  new kplus.HorizontalPodAutoscaler(chart, 'Hpa', {
    target: deployment,
    maxReplicas: 10,
    minReplicas: 2,
    metrics: [
      kplus.Metric.resource({
        name: 'cpu',
        target: kplus.MetricTarget.averageUtilization(50),
      }),
    ],
    scaleUp: {
      stabilizationWindow: Duration.minutes(5),
      strategy: kplus.ScalingStrategy.MAX_CHANGE,
      policies: [
        kplus.ScalingPolicy.pods(3, { scalingDuration: Duration.minutes(3) }),
        kplus.ScalingPolicy.percent(30),
      ],
    },
    scaleDown: {
      stabilizationWindow: Duration.minutes(5),
      strategy: kplus.ScalingStrategy.MAX_CHANGE,
      policies: [
        kplus.ScalingPolicy.pods(3, { scalingDuration: Duration.minutes(3) }),
        kplus.ScalingPolicy.percent(30),
      ],
    },
  });

  const manifest = Testing.synth(chart);
  expect(manifest).toMatchSnapshot();
  const spec = manifest[1].spec;
  expect(spec.maxReplicas).toEqual(10);
  expect(spec.minReplicas).toEqual(2);
  expect(spec.metrics).toEqual([{
    resource: {
      name: 'cpu',
      target: {
        averageUtilization: 50,
        type: 'Utilization',
      },
    },
    type: 'Resource',
  }]);
  expect(spec.behavior.scaleUp).toEqual({
    policies: [
      { periodSeconds: 180, type: 'Pods', value: 3 },
      { periodSeconds: 15, type: 'Percent', value: 30 },
    ],
    selectPolicy: 'Max',
    stabilizationWindowSeconds: 300,
  });
  expect(spec.behavior.scaleDown).toEqual({
    policies: [
      { periodSeconds: 180, type: 'Pods', value: 3 },
      { periodSeconds: 15, type: 'Percent', value: 30 },
    ],
    selectPolicy: 'Max',
    stabilizationWindowSeconds: 300,
  });
});
