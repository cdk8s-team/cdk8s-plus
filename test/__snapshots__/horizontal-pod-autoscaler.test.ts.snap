// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates HPA with expected spec, when metrics, scaleUp, and scaleDown are all configured 1`] = `
Array [
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "name": "test-deployment-c898c72d",
    },
    "spec": Object {
      "minReadySeconds": 0,
      "progressDeadlineSeconds": 600,
      "replicas": 2,
      "selector": Object {
        "matchLabels": Object {
          "cdk8s.io/metadata.addr": "test-Deployment-c83f5e59",
        },
      },
      "strategy": Object {
        "rollingUpdate": Object {
          "maxSurge": "25%",
          "maxUnavailable": "25%",
        },
        "type": "RollingUpdate",
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "cdk8s.io/metadata.addr": "test-Deployment-c83f5e59",
          },
        },
        "spec": Object {
          "automountServiceAccountToken": false,
          "containers": Array [
            Object {
              "image": "pod",
              "imagePullPolicy": "Always",
              "name": "main",
              "resources": Object {
                "limits": Object {
                  "cpu": "1500m",
                  "memory": "2048Mi",
                },
                "requests": Object {
                  "cpu": "1000m",
                  "memory": "512Mi",
                },
              },
              "securityContext": Object {
                "allowPrivilegeEscalation": false,
                "privileged": false,
                "readOnlyRootFilesystem": true,
                "runAsGroup": 26000,
                "runAsNonRoot": true,
                "runAsUser": 25000,
              },
            },
          ],
          "dnsPolicy": "ClusterFirst",
          "restartPolicy": "Always",
          "securityContext": Object {
            "fsGroupChangePolicy": "Always",
            "runAsNonRoot": true,
          },
          "setHostnameAsFQDN": false,
        },
      },
    },
  },
  Object {
    "apiVersion": "autoscaling/v2",
    "kind": "HorizontalPodAutoscaler",
    "metadata": Object {
      "name": "test-hpa-c8e6eb3e",
    },
    "spec": Object {
      "behavior": Object {
        "scaleDown": Object {
          "policies": Array [
            Object {
              "periodSeconds": 180,
              "type": "Pods",
              "value": 3,
            },
            Object {
              "periodSeconds": 15,
              "type": "Percent",
              "value": 30,
            },
          ],
          "selectPolicy": "Max",
          "stabilizationWindowSeconds": 300,
        },
        "scaleUp": Object {
          "policies": Array [
            Object {
              "periodSeconds": 180,
              "type": "Pods",
              "value": 3,
            },
            Object {
              "periodSeconds": 15,
              "type": "Percent",
              "value": 30,
            },
          ],
          "selectPolicy": "Max",
          "stabilizationWindowSeconds": 300,
        },
      },
      "maxReplicas": 10,
      "metrics": Array [
        Object {
          "resource": Object {
            "name": "cpu",
            "target": Object {
              "averageUtilization": 50,
              "type": "Utilization",
            },
          },
          "type": "Resource",
        },
      ],
      "minReplicas": 2,
      "scaleTargetRef": Object {
        "apiVersion": "apps/v1",
        "kind": "Deployment",
        "name": "test-deployment-c898c72d",
      },
    },
  },
]
`;

exports[`creates HPA, when metrics are not provided and one of the two target containers does not have any resources limits/requests 1`] = `
Array [
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "name": "test-deployment-c898c72d",
    },
    "spec": Object {
      "minReadySeconds": 0,
      "progressDeadlineSeconds": 600,
      "replicas": 2,
      "selector": Object {
        "matchLabels": Object {
          "cdk8s.io/metadata.addr": "test-Deployment-c83f5e59",
        },
      },
      "strategy": Object {
        "rollingUpdate": Object {
          "maxSurge": "25%",
          "maxUnavailable": "25%",
        },
        "type": "RollingUpdate",
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "cdk8s.io/metadata.addr": "test-Deployment-c83f5e59",
          },
        },
        "spec": Object {
          "automountServiceAccountToken": false,
          "containers": Array [
            Object {
              "image": "pod1",
              "imagePullPolicy": "Always",
              "name": "main",
              "securityContext": Object {
                "allowPrivilegeEscalation": false,
                "privileged": false,
                "readOnlyRootFilesystem": true,
                "runAsGroup": 26000,
                "runAsNonRoot": true,
                "runAsUser": 25000,
              },
            },
            Object {
              "image": "pod2",
              "imagePullPolicy": "Always",
              "name": "main",
              "resources": Object {
                "requests": Object {
                  "cpu": "256m",
                },
              },
              "securityContext": Object {
                "allowPrivilegeEscalation": false,
                "privileged": false,
                "readOnlyRootFilesystem": true,
                "runAsGroup": 26000,
                "runAsNonRoot": true,
                "runAsUser": 25000,
              },
            },
          ],
          "dnsPolicy": "ClusterFirst",
          "restartPolicy": "Always",
          "securityContext": Object {
            "fsGroupChangePolicy": "Always",
            "runAsNonRoot": true,
          },
          "setHostnameAsFQDN": false,
        },
      },
    },
  },
  Object {
    "apiVersion": "autoscaling/v2",
    "kind": "HorizontalPodAutoscaler",
    "metadata": Object {
      "name": "test-hpa-c8e6eb3e",
    },
    "spec": Object {
      "behavior": Object {
        "scaleDown": Object {
          "policies": Array [
            Object {
              "periodSeconds": 300,
              "type": "Pods",
              "value": 1,
            },
          ],
          "selectPolicy": "Max",
          "stabilizationWindowSeconds": 300,
        },
        "scaleUp": Object {
          "policies": Array [
            Object {
              "periodSeconds": 60,
              "type": "Pods",
              "value": 4,
            },
            Object {
              "periodSeconds": 60,
              "type": "Percent",
              "value": 200,
            },
          ],
          "selectPolicy": "Max",
          "stabilizationWindowSeconds": 0,
        },
      },
      "maxReplicas": 10,
      "minReplicas": 1,
      "scaleTargetRef": Object {
        "apiVersion": "apps/v1",
        "kind": "Deployment",
        "name": "test-deployment-c898c72d",
      },
    },
  },
]
`;

exports[`creates HPA, when minReplicas is same as maxReplicas 1`] = `
Array [
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "name": "test-deployment-c898c72d",
    },
    "spec": Object {
      "minReadySeconds": 0,
      "progressDeadlineSeconds": 600,
      "replicas": 2,
      "selector": Object {
        "matchLabels": Object {
          "cdk8s.io/metadata.addr": "test-Deployment-c83f5e59",
        },
      },
      "strategy": Object {
        "rollingUpdate": Object {
          "maxSurge": "25%",
          "maxUnavailable": "25%",
        },
        "type": "RollingUpdate",
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "cdk8s.io/metadata.addr": "test-Deployment-c83f5e59",
          },
        },
        "spec": Object {
          "automountServiceAccountToken": false,
          "containers": Array [
            Object {
              "image": "pod",
              "imagePullPolicy": "Always",
              "name": "main",
              "resources": Object {
                "limits": Object {
                  "cpu": "1500m",
                  "memory": "2048Mi",
                },
                "requests": Object {
                  "cpu": "1000m",
                  "memory": "512Mi",
                },
              },
              "securityContext": Object {
                "allowPrivilegeEscalation": false,
                "privileged": false,
                "readOnlyRootFilesystem": true,
                "runAsGroup": 26000,
                "runAsNonRoot": true,
                "runAsUser": 25000,
              },
            },
          ],
          "dnsPolicy": "ClusterFirst",
          "restartPolicy": "Always",
          "securityContext": Object {
            "fsGroupChangePolicy": "Always",
            "runAsNonRoot": true,
          },
          "setHostnameAsFQDN": false,
        },
      },
    },
  },
  Object {
    "apiVersion": "autoscaling/v2",
    "kind": "HorizontalPodAutoscaler",
    "metadata": Object {
      "name": "test-hpa-c8e6eb3e",
    },
    "spec": Object {
      "behavior": Object {
        "scaleDown": Object {
          "policies": Array [
            Object {
              "periodSeconds": 300,
              "type": "Pods",
              "value": 1,
            },
          ],
          "selectPolicy": "Max",
          "stabilizationWindowSeconds": 300,
        },
        "scaleUp": Object {
          "policies": Array [
            Object {
              "periodSeconds": 60,
              "type": "Pods",
              "value": 4,
            },
            Object {
              "periodSeconds": 60,
              "type": "Percent",
              "value": 200,
            },
          ],
          "selectPolicy": "Max",
          "stabilizationWindowSeconds": 0,
        },
      },
      "maxReplicas": 10,
      "minReplicas": 10,
      "scaleTargetRef": Object {
        "apiVersion": "apps/v1",
        "kind": "Deployment",
        "name": "test-deployment-c898c72d",
      },
    },
  },
]
`;

exports[`creates HPA, when target is a StatefulSet 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "Service",
    "metadata": Object {
      "name": "test-testservice-c876eb6f",
    },
    "spec": Object {
      "externalIPs": Array [],
      "ports": Array [
        Object {
          "port": 80,
        },
      ],
      "selector": Object {
        "cdk8s.io/metadata.addr": "test-StatefulSet-c809b559",
      },
      "type": "ClusterIP",
    },
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "StatefulSet",
    "metadata": Object {
      "name": "test-statefulset-c8a6ec86",
    },
    "spec": Object {
      "minReadySeconds": 0,
      "podManagementPolicy": "OrderedReady",
      "replicas": 1,
      "selector": Object {},
      "serviceName": "test-testservice-c876eb6f",
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "cdk8s.io/metadata.addr": "test-StatefulSet-c809b559",
          },
        },
        "spec": Object {
          "automountServiceAccountToken": false,
          "containers": Array [
            Object {
              "image": "foobar",
              "imagePullPolicy": "Always",
              "name": "main",
              "resources": Object {
                "limits": Object {
                  "cpu": "1500m",
                  "memory": "2048Mi",
                },
                "requests": Object {
                  "cpu": "1000m",
                  "memory": "512Mi",
                },
              },
              "securityContext": Object {
                "allowPrivilegeEscalation": false,
                "privileged": false,
                "readOnlyRootFilesystem": true,
                "runAsGroup": 26000,
                "runAsNonRoot": true,
                "runAsUser": 25000,
              },
            },
          ],
          "dnsPolicy": "ClusterFirst",
          "restartPolicy": "Always",
          "securityContext": Object {
            "fsGroupChangePolicy": "Always",
            "runAsNonRoot": true,
          },
          "setHostnameAsFQDN": false,
        },
      },
      "updateStrategy": Object {
        "rollingUpdate": Object {
          "partition": 0,
        },
        "type": "RollingUpdate",
      },
    },
  },
  Object {
    "apiVersion": "autoscaling/v2",
    "kind": "HorizontalPodAutoscaler",
    "metadata": Object {
      "name": "test-hpa-c8e6eb3e",
    },
    "spec": Object {
      "behavior": Object {
        "scaleDown": Object {
          "policies": Array [
            Object {
              "periodSeconds": 300,
              "type": "Pods",
              "value": 1,
            },
          ],
          "selectPolicy": "Max",
          "stabilizationWindowSeconds": 300,
        },
        "scaleUp": Object {
          "policies": Array [
            Object {
              "periodSeconds": 60,
              "type": "Pods",
              "value": 4,
            },
            Object {
              "periodSeconds": 60,
              "type": "Percent",
              "value": 200,
            },
          ],
          "selectPolicy": "Max",
          "stabilizationWindowSeconds": 0,
        },
      },
      "maxReplicas": 10,
      "minReplicas": 1,
      "scaleTargetRef": Object {
        "apiVersion": "apps/v1",
        "kind": "StatefulSet",
        "name": "test-statefulset-c8a6ec86",
      },
    },
  },
]
`;

exports[`creates HPA, when target is a deployment 1`] = `
Array [
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "name": "test-deployment-c898c72d",
    },
    "spec": Object {
      "minReadySeconds": 0,
      "progressDeadlineSeconds": 600,
      "replicas": 2,
      "selector": Object {
        "matchLabels": Object {
          "cdk8s.io/metadata.addr": "test-Deployment-c83f5e59",
        },
      },
      "strategy": Object {
        "rollingUpdate": Object {
          "maxSurge": "25%",
          "maxUnavailable": "25%",
        },
        "type": "RollingUpdate",
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "cdk8s.io/metadata.addr": "test-Deployment-c83f5e59",
          },
        },
        "spec": Object {
          "automountServiceAccountToken": false,
          "containers": Array [
            Object {
              "image": "pod",
              "imagePullPolicy": "Always",
              "name": "main",
              "resources": Object {
                "limits": Object {
                  "cpu": "1500m",
                  "memory": "2048Mi",
                },
                "requests": Object {
                  "cpu": "1000m",
                  "memory": "512Mi",
                },
              },
              "securityContext": Object {
                "allowPrivilegeEscalation": false,
                "privileged": false,
                "readOnlyRootFilesystem": true,
                "runAsGroup": 26000,
                "runAsNonRoot": true,
                "runAsUser": 25000,
              },
            },
          ],
          "dnsPolicy": "ClusterFirst",
          "restartPolicy": "Always",
          "securityContext": Object {
            "fsGroupChangePolicy": "Always",
            "runAsNonRoot": true,
          },
          "setHostnameAsFQDN": false,
        },
      },
    },
  },
  Object {
    "apiVersion": "autoscaling/v2",
    "kind": "HorizontalPodAutoscaler",
    "metadata": Object {
      "name": "test-hpa-c8e6eb3e",
    },
    "spec": Object {
      "behavior": Object {
        "scaleDown": Object {
          "policies": Array [
            Object {
              "periodSeconds": 300,
              "type": "Pods",
              "value": 1,
            },
          ],
          "selectPolicy": "Max",
          "stabilizationWindowSeconds": 300,
        },
        "scaleUp": Object {
          "policies": Array [
            Object {
              "periodSeconds": 60,
              "type": "Pods",
              "value": 4,
            },
            Object {
              "periodSeconds": 60,
              "type": "Percent",
              "value": 200,
            },
          ],
          "selectPolicy": "Max",
          "stabilizationWindowSeconds": 0,
        },
      },
      "maxReplicas": 10,
      "minReplicas": 1,
      "scaleTargetRef": Object {
        "apiVersion": "apps/v1",
        "kind": "Deployment",
        "name": "test-deployment-c898c72d",
      },
    },
  },
]
`;
